### PR TITLE
gccCrossStageStatic: enable dynamic libraries, rename it

### LINKED ISSUE #40797: libgcc_s.so is missing from glibc when cross compiling
## Issue description

I'm not sure what the intended behavior actually is, but there is an issue with `libgcc_s.so` when cross-compiling, or perhaps the native case is actually incorrect.

I came across this problem when debugging why `tmon` crashes with the error: `libgcc_s.so.1 must be installed for pthread_cancel to work` when exiting, but only when it is cross-compiled. I noticed that, for native builds only, glibc [copies libgcc_s.so from bootstrap gcc](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/glibc/default.nix#L43). This  `libgcc_s.so` from the bootstrap tools is present in the glibc used to build all other derivations, which does not seem right to me (it doesn't seem like normal derivations should have access to any libraries/binaries from the the bootstrap tools).

pthreads needs `libgcc_s.so`, and it is able to find it when built natively because `libgcc_s.so` is part of glibc, which is on the RPATH. When cross-compiling, glibc does not have `libgcc_s.so`, and the gcc libs are not on the RPATH, so pthreads cannot find `libgcc_s.so`.

I can see that there are a number of cases where people have worked around this problem in nixpkgs, by manually linking to `libgcc_s`. This solves the problem for tmon, but I still think there is a bug because of the different behavior between native and cross builds.

cc @vcunat

### LINKED ISSUE #240484: cross compiled binaries segfault on pthread_cancel(3) due to missing libgcc_s.so
This is entirely my fault. :cry: 

##### The problem

Any binary using our `pkgsCross.*.glibc` built after #209870 but before #238154 which attempts to use `pthread_cancel(3)` will (deliberately) segfault.

Since there is no error message and because `pthread_cancel(3)` is a somewhat-obscure API this is extremely frustrating to troubleshoot.

##### The reason for the problem

The cross compiler we were using to build `glibc` between those two commits didn't have a `libgcc_s.so`.  So when `glibc` attempts to `dlopen()` `libgcc_s.so`, it can't find it.  The `glibc` developers believe that segfaulting without an error message is the right thing to do in this situation.

##### Additional details

Our cross compilers don't (currently) bootstrap using the normal stdenv stages.

The "first stage" gcc (`crossStageStatic` for cross), doesn't produce a `libgcc_s.so` since it is a "static only" compiler.  This compiler is used to compile `glibc`.  Because `libgcc_s.so` doesn't exist when `glibc` is built, `pkgsCross.*.glibc.stdenv.cc.cc.libgcc` doesn't exist.  So `user-defined-trusted-dirs` isn't added to `Makeflags` and `glibc` doesn't know where to look for `libgcc_s.so`.  Also, cross-compiled `glibc`s won't pull `libgcc` into their closure like they should.

This is fixed by:

https://github.com/NixOS/nixpkgs/pull/238154

### PR DESCRIPTION
**Pull request by ghost**
Best reviewed one commit at a time.

##### Motivation

1. Fix a serious bug in our `pkgsCross` bootstrap: https://github.com/NixOS/nixpkgs/pull/240484.  This is the reason for the backport request.
   a. Also: Closes #40797

2. Make our cross-compiler bootstrap look more like our `stdenv` bootstrap, as a step towards unifying them.

##### Summary

This PR allows `gccCrossStageStatic` to build dynamically-linked libraries.  Since is no longer restricted to building static libraries its name is no longer appropriate, and this PR also renames it to the more-accurate `gccWithoutTargetLibc`.

##### Detailed description of changes

By default, you can't build a gcc that knows how to create dynamic libraries unless you have already built the targetPlatform libc.

Because of this, our gcc cross-compiler is built in three stages:

  1. Build a cross-compiler (gccCrossStageStatic) that can build         only static libraries.

  2. Use gccCrossStageStatic to compile the targetPlatform `libc.a`.

  5. Use the targetPlatform `libc.a` to build a fully-capable cross         compiler.

You might notice that this pattern looks very similar to what we do with `xgcc` and `glibc` in the stdenv bootstrap.  Indeed it is!  I would like to work towards getting the existing stdenv bootstrap to handle cross compilers as well.  However we don't want to cripple `stdenv.xgcc` by taking away its ability to build dynamic libraries.

It turns out that the only thing gcc needs the targetPlatform libc for is to emit a DT_NEEDED for `-lc` into `libgcc.so`.  That's it! And since we don't use `gccCrossStageStatic` to build anything other than libc, it's safe to omit the `DT_NEEDED` because that `libgcc` will never be loaded by anything other than `libc`.  So `libc` will already be in the process's address space.

Other people have noticed this; crosstool-ng has been using this approach for a very long time:

https://github.com/crosstool-ng/crosstool-ng/blob/36ad0b17a732aaffe4701d5d8d410d6e3e3abba9/scripts/build/cc/gcc.sh#L638-L640

Closes #240484

Closes #240435

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] powerpc64le-linux
  - [x] aarch64-linux (cross from x86_64-linux)
  - [x] mips64el-linuxabi64 (cross from x86_64-linux)
  - [x] mips64el-linuxabin32 (cross from x86_64-linux)
  - [x] arm-eabi `gcc6` (cross from x86_64-linux)
- Built to test gcc versions
  - `pkgsCross.aarch64-multiplatform`
  - [x] `.pkgsBuildTarget.gcc48`
  - [x] `.pkgsBuildTarget.gcc49`
  - [x] `.pkgsBuildTarget.gcc6`
  - [x] `.pkgsBuildTarget.gcc7`
  - [x] `.pkgsBuildTarget.gcc8`
  - [x] `.pkgsBuildTarget.gcc9`
  - [x] `.pkgsBuildTarget.gcc10`
  - [x] `.pkgsBuildTarget.gcc11`
  - [x] `.pkgsBuildTarget.gcc12`
  - [x] `.pkgsBuildTarget.gcc13`
- Tested complete userspace rebuild on platform(s):
  - [x] powerpc64le-linux
  - [x] aarch64-linux (cross from x86_64-linux)
  - [x] mips64el-linuxabi64+mips64el-linuxabin32 (mixed) (cross from x86_64-linux)

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
I'm testing builds on a pretty large matrix of target platforms and gcc versions.

In spite of this, I have the nagging feeling that this is going to break some obscure gcc version on an infrequently-used platform.  I've tried to keep this PR `master`-eligible (i.e. no mass-rebuilds) so that if that kind of breakage does happen,

a. People will be able to find the cause (this PR) without having to `git bisect` across a merge of `staging` into `master`
b. The fix PR can go to `master` instead of `staging`.

**Comment by amjoseph-nixpkgs** [comment]
This doesn't work on powerpc64le.  Seems to work everywhere else.  Investigating.

**Comment by amjoseph-nixpkgs** [comment]
Everything tests fine, rebooted the laptop with a complete system (bootloader, kernel, userspace) built this way.

However the change made in gcc commit [e50084fa44cb68c447efc11e53ec16bf09a578c0](https://gcc.gnu.org/git/?p=gcc.git;a=blobdiff;f=gcc/configure.in;h=824008e4e69ff2e55a232963dd0e519a3cb7c227;hp=6360c2488ba2713fd1bcabf1cd2e117f84d72e3b;hb=e50084fa44cb68c447efc11e53ec16bf09a578c0;hpb=b348267b2e2243e76433b8b7798695787a80cd0e) needs to be made to `libstdc++` as well; I will send a patch to do that upstream and include it here.

**Comment by Ericson2314** [comment]
Looks good so far, @amjoseph-nixpkgs

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.glibc.passthru.libgcc

**Comment by amjoseph-nixpkgs** [comment]
>  pkgsCross.aarch64-multiplatform.glibc.passthru.libgcc on x86_64-linux — Success
[Details](https://github.com/NixOS/nixpkgs/pull/238154/checks?check_run_id=14643193782)

Sweet.  Prior to this PR that didn't even evaluate.

**Comment by amjoseph-nixpkgs** [comment]
Split out https://github.com/NixOS/nixpkgs/pull/240575

**Comment by amjoseph-nixpkgs** [comment]
At latest push 9059d97ffddf306b65697a7ff95d697ade9a0c0a, `nix-instantiate . -A pkgsStatic.busyBox` gives the same derivation before and after this PR.

I'll undraftify in the morning once my rebuilds have finished.

**Comment by trofi** [comment]
```nix
+ lib.optionalString (targetPlatform != hostPlatform && withoutTargetLibc && enableShared)
  (lib.optionalString (!stdenv.targetPlatform.isPower) ''
    echo 'libgcc.a: crti.o crtn.o' >> libgcc/Makefile.in
  '' + ''
    echo 'SHLIB_LC=' >> libgcc/Makefile.in
  '')
```

It's a very scary hack to change `libgcc`. Whould be nice to have it upstreamed in some form.

**Comment by amjoseph-nixpkgs** [comment]
> It's a very scary hack to change `libgcc`.

Note: this doesn't change `libgcc`'s code, just its build process -- exactly what we are already doing (to *every* package) in `cc-wrapper` and `ld-wrapper`!  People have been using the exact `Makefile` rule above [for over twenty years](https://web.archive.org/web/20170224235700/http://frank.harvard.edu/~coldwell/toolchain/t-linux.diff) to cross-compile gcc without an extra recompile.  It has worked since gcc 3.3!

> Whould be nice to have it upstreamed in some form.

Agree, but that's not going to happen.  AFAICT upstream considers the circular dependency between `gcc` and `glibc` to be a "feature not a bug".  From their perspective circular dependencies are okay and the fact that Nix disallows them is a bug in Nix.

**Review by Mindavi — APPROVED** [review]
The hasSharedLibraries seems a bit weird, but the changes look sane and I trust this is verified very thoroughly

**Comment by trofi** [comment]
Bisect says 2affd455a40a28825f356307ce5bd8fa2f202217 `gccCrossStageStatic: enable dynamic libraries, rename to gccWithoutTargetLibc` broke `pkgsCross.mingwW64.stdenv` target:

```
$ nix build --no-link github:NixOS/nixpkgs#pkgsCross.mingwW64.stdenv -L
...
x86_64-w64-mingw32-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/i386/crti.S: Assembler messages:
x86_64-w64-mingw32-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/i386/crti.S:34: Warning: .type pseudo-op used outside of .def/.endef: ignored.
x86_64-w64-mingw32-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/i386/crti.S:34: Error: junk at end of line, first unrecognized character is `_'
x86_64-w64-mingw32-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/i386/crti.S:39: Warning: .type pseudo-op used outside of .def/.endef: ignored.
x86_64-w64-mingw32-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/i386/crti.S:39: Error: junk at end of line, first unrecognized character is `_'
x86_64-w64-mingw32-stage-static-gcc> make[1]: *** [Makefile:1079: crti.o] Error 1
x86_64-w64-mingw32-stage-static-gcc> make[1]: *** Waiting for unfinished jobs....
```

**Comment by trofi** [comment]
Result of `nixpkgs-review pr 238154` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>50 packages failed to build:</summary>
  <ul>
    <li>dxvk</li>
    <li>dxvk.bin</li>
    <li>dxvk.lib</li>
    <li>exactaudiocopy</li>
    <li>grapejuice</li>
    <li>grapejuice.dist</li>
    <li>klipper-firmware</li>
    <li>klipper-flash</li>
    <li>pipelight</li>
    <li>pkgsLLVM.stdenv</li>
    <li>playonlinux</li>
    <li>q4wine</li>
    <li>vkd3d</li>
    <li>vkd3d-proton</li>
    <li>winbox</li>
    <li>wine (winePackages.full ,winePackages.stableFull)</li>
    <li>wine-staging (winePackages.stagingFull)</li>
    <li>wine-wayland (winePackages.waylandFull)</li>
    <li>wine64 (wine64Packages.full ,wine64Packages.stableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable)</li>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.unstable</li>
    <li>wine64Packages.unstableFull</li>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
    <li>winePackages.base (winePackages.stable)</li>
    <li>winePackages.staging</li>
    <li>winePackages.unstable</li>
    <li>winePackages.unstableFull</li>
    <li>winePackages.wayland</li>
    <li>wineWow64Packages.base (wineWow64Packages.stable)</li>
    <li>wineWow64Packages.full (wineWow64Packages.stableFull)</li>
    <li>wineWow64Packages.staging</li>
    <li>wineWow64Packages.stagingFull</li>
    <li>wineWow64Packages.unstable</li>
    <li>wineWow64Packages.unstableFull</li>
    <li>wineWow64Packages.wayland</li>
    <li>wineWow64Packages.waylandFull</li>
    <li>wineWowPackages.base (wineWowPackages.stable)</li>
    <li>wineWowPackages.full (wineWowPackages.stableFull)</li>
    <li>wineWowPackages.staging</li>
    <li>wineWowPackages.stagingFull</li>
    <li>wineWowPackages.unstable</li>
    <li>wineWowPackages.unstableFull</li>
    <li>wineWowPackages.wayland</li>
    <li>wineWowPackages.waylandFull</li>
    <li>wineasio</li>
    <li>yabridge</li>
    <li>yabridgectl</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>armTrustedFirmwareTools</li>
    <li>fusee-launcher</li>
    <li>gnuk</li>
    <li>klipper-genconf</li>
    <li>qmk</li>
    <li>qmk.dist</li>
    <li>simavr</li>
    <li>spike</li>
    <li>tests.makeBinaryWrapper</li>
    <li>tinygo</li>
  </ul>
</details>

**Comment by amjoseph-nixpkgs** [comment]
> Result of `nixpkgs-review pr 238154` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

https://github.com/NixOS/nixpkgs/pull/243230 should fix most of this.  Testing now.

**Comment by trofi** [comment]
Bisect claims that one of these commits also broke `pkgsCross.m68k.stdenv` in `master`:

```
$ nix build --no-link -f. pkgsCross.m68k.stdenv -L
...
m68k-unknown-linux-gnu-stage-static-gcc> make[1]: Leaving directory '/build/build/m68k-unknown-linux-gnu/libgcc'
m68k-unknown-linux-gnu-stage-static-gcc> Moving /nix/store/8658fc2zsh062ibn1l4vsvsapy4jppkz-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0/m68k-unknown-linux-gnu/lib/libgcc_s.so to /nix/store/sf4g7jbpm502pp8lr2jh0qvxq2wqv0db-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0-lib/m68k-unknown-linux-gnu/lib/libgcc_s.so
m68k-unknown-linux-gnu-stage-static-gcc> Moving /nix/store/8658fc2zsh062ibn1l4vsvsapy4jppkz-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0/m68k-unknown-linux-gnu/lib/libgcc_s.so.2 to /nix/store/sf4g7jbpm502pp8lr2jh0qvxq2wqv0db-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0-lib/m68k-unknown-linux-gnu/lib/libgcc_s.so.2
m68k-unknown-linux-gnu-stage-static-gcc> Removing empty /nix/store/8658fc2zsh062ibn1l4vsvsapy4jppkz-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0/m68k-unknown-linux-gnu/lib/ and (possibly) its parents
m68k-unknown-linux-gnu-stage-static-gcc> /nix/store/qr60k7sz61hvv2baadlfv5qjr8q7vlg6-builder.sh: line 243: type: install_name_tool: not found
m68k-unknown-linux-gnu-stage-static-gcc> preFixupLibGccPhase
m68k-unknown-linux-gnu-stage-static-gcc> mv: cannot stat '/nix/store/sf4g7jbpm502pp8lr2jh0qvxq2wqv0db-m68k-unknown-linux-gnu-stage-static-gcc-12.3.0-lib/m68k-unknown-linux-gnu/lib/libgcc_s.so.1': No such file or directory
```

I think `m68k` defines `libgcc_s.so.2`, not `libgcc_s.so.1`.

**Comment by trofi** [comment]
And bisect claims that one of these commits also broke `pkgsCross.s390.stdenv` in master:

```
$ nix build --no-link -f. pkgsCross.s390.stdenv -L
...
s390-unknown-linux-gnu-stage-static-gcc> /build/build/./gcc/xgcc -B/build/build/./gcc/    -O2  -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include  -fPIC -mlong-double-128 -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector -Dinhibit_libc -fPIC -mlong-double-128 -I. -I. -I../.././gcc -I../../../gcc-12.3.0/libgcc -I../../../gcc-12.3.0/libgcc/. -I../../../gcc-12.3.0/libgcc/../gcc -I../../../gcc-12.3.0/libgcc/../include  -DHAVE_CC_TLS   -o morestack_s.o -MT morestack_s.o -MD -MP -MF morestack_s.dep -DSHARED -c -xassembler-with-cpp ../../../gcc-12.3.0/libgcc/config/s390/morestack.S
s390-unknown-linux-gnu-stage-static-gcc> make[1]: *** No rule to make target '../../../gcc-12.3.0/libgcc/config/s390/crti.S', needed by 'crti.o'.  Stop.
s390-unknown-linux-gnu-stage-static-gcc> make[1]: *** Waiting for unfinished jobs....
s390-unknown-linux-gnu-stage-static-gcc> /nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash ../../../gcc-12.3.0/libgcc/../move-if-change tmp-libgcc_tm.h libgcc_tm.h
s390-unknown-linux-gnu-stage-static-gcc> echo timestamp > libgcc_tm.stamp
s390-unknown-linux-gnu-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/s390/morestack.S: Assembler messages:
s390-unknown-linux-gnu-stage-static-gcc> ../../../gcc-12.3.0/libgcc/config/s390/morestack.S:600: Warning: ignoring incorrect section type for .init_array.00000
s390-unknown-linux-gnu-stage-static-gcc> make[1]: Leaving directory '/build/build/s390-unknown-linux-gnu/libgcc'
s390-unknown-linux-gnu-stage-static-gcc> make: *** [Makefile:12947: all-target-libgcc] Error 2
error: builder for '/nix/store/lffjd90i8pa7mbwjcxpwf8cxaa2xqljp-s390-unknown-linux-gnu-stage-static-gcc-12.3.0.drv' failed with exit code 2;
```

Don't see why immediately.

**Comment by amjoseph-nixpkgs** [comment]
> Bisect claims that one of these commits also broke `pkgsCross.m68k.stdenv` in `master`:

> And bisect claims that one of these commits also broke `pkgsCross.s390.stdenv` in master:

Please add these to `tests.cross.sanity` so they are tested for; if there are no tests for something, it is certain to break.

Also, I'm pretty sure none of us can afford S390 hardware.

**Comment by trofi** [comment]
```
$ nix build -f.  tests.cross.sanity
error: attribute 'sanity' in selection path 'tests.cross.sanity' not found
```

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build tests.cross.sanity

**Comment by trofi** [comment]
Added `assert` against an attrset field caused the non-determinism regression: https://github.com/NixOS/nixpkgs/issues/244045 (a patch for `nix` to ease reproducibility: https://github.com/NixOS/nix/pull/8711)

**Comment by trofi** [comment]
Another regression is `pkgCross.loongarch64-linux.stdenv` reported by jiegec: https://github.com/NixOS/nixpkgs/issues/252590

**Comment by mikatammi** [comment]
I'm guessing the automatic backport failed and we never got backport to 23.05? Could this be backported, it would be nice, as I'm seeing the cross-compiled Wayland failing on runtime, and this PR fixes the issue

**Comment by alyssais** [comment]
This PR caused some regressions, so any potential backport should probably be okayed by the release managers before being merged.

**Pull request merged**
